README.md
# ECUALEAD Message API

Utility functions for ECUALEAD Message API integration. This library provide the functions to trigger mail notifications to be delivered to users.

The package is developed with the aim of being used in conjunction with the rest of the packages of the platform, but it don't restrict use it as standalone package. The request validation is performed against the Identity Management Service.

[![Version npm](https://img.shields.io/npm/v/@ecualead/msg.svg?style=flat-square)](https://www.npmjs.com/package/@ecualead/msg)[![npm Downloads](https://img.shields.io/npm/dm/@ecualead/msg.svg?style=flat-square)](https://npmcharts.com/compare/@ecualead/msg?minimal=true)[![Build Status](https://gitlab.com/ecualead/msg/badges/master/pipeline.svg)](https://gitlab.com/iecualead/msg)[![coverage testing report](https://gitlab.com/ecualead/msg/badges/master/coverage.svg)](https://gitlab.com/ecualead/msg/-/commits/master)

[![NPM](https://nodei.co/npm/@ecualead/msg.png?downloads=true&downloadRank=true)](https://nodei.co/npm/@ecualead/msg/)

## Installation

```bash
npm install @ecualead/msg
```

## Send mail notifications

To send mail notifications first the utility must be configured with the mailer server and the access token to authenticate the request. Access token must belong to api service, users don't have allowed to send mail notifications. The used api service must be authorized to send mails.

```js
import { MailCtrl } from "@ecualead/msg";
MailCtrl.setup("https://myserver.com", "myaccesstoken");
```

Once the api is configured we can send mail notifications using

```js
MailCtrl.send(subject: string, body?: string, template?:string, data?: any, to?: string | string[], cc?: string | string[], bcc?: string | string[], attachments?: IMailAttachment[]): Promise<void>
```

Where:

- `subject`: Subject of the mail notification.
- `body`: Body of the mail message to be sent.
- `template`: The mail notification to be triggered. This identifies the mail template to be used on the notification.
- `data`: Additional data to replace in the mail template.
- `to`: List of receipts to send the notification.
- `cc`: List of receipts to send the notification as copy.
- `bcc`: List of receipts to send the notification as hidden copy.
- `attachments`: List of base64 files to be included as attachments to the mail.

There are some things to keep in mind:

- Parameters `body` and` template` are exclusive, with higher precedence for `body`. One of them must be set, otherwise an error will be thrown.
- In the case of `to`,` cc` or `bcc`, at least one of them must be set, otherwise an error will occur.

Attachments are defined as

```js
interface IMailAttachment {
  filename: string;
  content: string;
  encoding?: string;
}
```

Attachment can be loaded using

```js
MailCtrl.attachmentFromFile(filepath: string, filename?: string): IMailAttachment;
```


CHANGELOG.md
# CHANGELOG

## [2.0.1] - 2022-01-10

- Add function to load attachment from file
- Add auth dependency

## [2.0.0] - 2022-01-07

- Update package namespace
- Add support to send mail attachments
- Update dependencies versions

## [1.0.0] - 2021-10-21

- Update package version

## [1.0.0] - 2021-06-27

- Initial mailer version
- Support to send mail notifications
- Support for mail templates
- Support for mail server settings


tests/mail.test.ts
/**
 * Copyright (C) 2020 - 2021 IKOA Business Opportunity
 *
 * All Rights Reserved
 * Author: Reinier Millo Sánchez <millo@ikoabo.com>
 *
 * This file is part of the IKOA Business Oportunity Mailer API
 * It can't be copied and/or distributed without the express
 * permission of the author.
 */
import "mocha";
import chai from "chai";
import { MailCtrl, ERRORS } from '../src';
import { HTTP_STATUS, SERVER_ERRORS } from "@ikoabo/core";
const expect = chai.expect;

describe("Mail sender", () => {
  it("Invalid mail server", () => {
    MailCtrl.setup("", "token");
    return MailCtrl.send("any", "test mail", "body", "template", null, "test@mail.com")
      .then(() => { throw new Error('was not supposed to succeed'); })
      .catch((err) => {
        expect(err.boError).to.equal(ERRORS.INVALID_MAIL_SERVER);
        expect(err.boStatus).to.equal(HTTP_STATUS.HTTP_4XX_NOT_ACCEPTABLE);
      })
  });

  it("Invalid mail receiver", async () => {
    MailCtrl.setup("server", "token");
    return MailCtrl.send("any", "test mail", "body", "template")
      .then(() => { throw new Error('was not supposed to succeed'); })
      .catch((err) => {
        expect(err.boError).to.equal(ERRORS.INVALID_MAIL_DATA);
        expect(err.boStatus).to.equal(HTTP_STATUS.HTTP_4XX_NOT_ACCEPTABLE);
      })
  });

  it("Invalid mail body", async () => {
    MailCtrl.setup("server", "token");
    return MailCtrl.send("any", "test mail", null, null, null, "sample@mail.com")
      .then(() => { throw new Error('was not supposed to succeed'); })
      .catch((err) => {
        expect(err.boError).to.equal(ERRORS.INVALID_MAIL_DATA);
        expect(err.boStatus).to.equal(HTTP_STATUS.HTTP_4XX_NOT_ACCEPTABLE);
      })
  });

  it("Unknown server", async () => {
    MailCtrl.setup("https://invalid.ikoabo.com", "token");
    return MailCtrl.send("any", "test mail", "Test mail", null, null, "sample@mail.com")
      .then(() => { throw new Error('was not supposed to succeed'); })
      .catch((err) => {
        expect(err.boError).to.equal(ERRORS.UNKNOWN_MAIL_SERVER_ERROR);
        expect(err.boStatus).to.equal(HTTP_STATUS.HTTP_4XX_FORBIDDEN);
      })
  }).timeout(10000);

  it("Try send mail with body", async () => {
    MailCtrl.setup("https://mailer.ikoabo.com", "token");
    return MailCtrl.send("any", "test mail", "Test mail", null, null, "sample@mail.com")
      .then(() => { throw new Error('was not supposed to succeed'); })
      .catch((err) => {
        expect(err.boError).to.equal(SERVER_ERRORS.UNKNOWN_ERROR);
        expect(err.boStatus).to.equal(HTTP_STATUS.HTTP_4XX_BAD_REQUEST);
      })
  }).timeout(10000);

  it("Try send mail with template", async () => {
    MailCtrl.setup("https://mailer.ikoabo.com", "token");
    return MailCtrl.send("any", "test mail", null, "custom-template", null, "sample@mail.com")
      .then(() => { throw new Error('was not supposed to succeed'); })
      .catch((err) => {
        expect(err.boError).to.equal(SERVER_ERRORS.UNKNOWN_ERROR);
        expect(err.boStatus).to.equal(HTTP_STATUS.HTTP_4XX_BAD_REQUEST);
      })
  }).timeout(10000);
});


tests/errors.test.ts
/**
 * Copyright (C) 2020 - 2021 IKOA Business Opportunity
 *
 * All Rights Reserved
 * Author: Reinier Millo Sánchez <millo@ikoabo.com>
 *
 * This file is part of the IKOA Business Oportunity Mailer API
 * It can't be copied and/or distributed without the express
 * permission of the author.
 */
import "mocha";
import chai from "chai";
import { ERRORS } from '../src';
const expect = chai.expect;

describe("Testing errors definition", () => {
    it("Ammount of errors", (done) => {
        expect(Object.keys(ERRORS).length).to.be.equal(12);
        done();
    });

    it("General errors", (done) => {
        expect(ERRORS.UNKNOWN_MAIL_SERVER_ERROR.value).to.be.equal(1500);
        expect(ERRORS.UNKNOWN_MAIL_SERVER_ERROR.str).to.be.equal("unknown-mail-server-error");

        expect(ERRORS.NOT_AUTHORIZED.value).to.be.equal(1501);
        expect(ERRORS.NOT_AUTHORIZED.str).to.be.equal("not-authorized");

        done();
    });

    it("Mail errors", (done) => {
        expect(ERRORS.INVALID_MAIL_TEMPLATE.value).to.be.equal(1520);
        expect(ERRORS.INVALID_MAIL_TEMPLATE.str).to.be.equal("invalid-mail-template");

        expect(ERRORS.INVALID_MAIL_SETTINGS.value).to.be.equal(1521);
        expect(ERRORS.INVALID_MAIL_SETTINGS.str).to.be.equal("invalid-mail-settings");

        expect(ERRORS.INVALID_MAIL_SERVER.value).to.be.equal(1522);
        expect(ERRORS.INVALID_MAIL_SERVER.str).to.be.equal("invalid-mail-server");

        expect(ERRORS.INVALID_MAIL_DATA.value).to.be.equal(1523);
        expect(ERRORS.INVALID_MAIL_DATA.str).to.be.equal("invalid-mail-data");

        expect(ERRORS.MAIL_SERVER_ERROR.value).to.be.equal(1524);
        expect(ERRORS.MAIL_SERVER_ERROR.str).to.be.equal("mail-server-error");

        done();
    });
});


src/controllers/mailer.controller.ts
/**
 * Copyright (C) 2020-2022 ECUALEAD
 * All Rights Reserved
 * Author: Reinier Millo Sánchez <rmillo@ecualead.com>
 *
 * This file is part of the ECUALEAD Message Server API.
 * It can't be copied and/or distributed without the express
 * permission of the author.
 */
import { SERVER_ERRORS, HTTP_STATUS, Objects } from "@ecualead/server";
import { ERRORS } from "../constants/errors.enum";
import axios, { AxiosError, AxiosResponse } from "axios";
import fs from "fs";
import { basename } from "path";
import { AuthenticationCtrl } from "@ecualead/auth";

export interface IMailAttachment {
  filename: string;
  content: string;
  encoding?: string;
}

class Mails {
  private static _instance: Mails;
  private _service: string;

  /**
   * Private constructor to allow singleton instance
   */
  private constructor() {
    /* Do nothing on constructor */
  }

  /**
   * Retrieve singleton class instance
   */
  public static get shared(): Mails {
    if (!Mails._instance) {
      Mails._instance = new Mails();
    }
    return Mails._instance;
  }

  public setup(service: string) {
    this._service = service;
  }

  public send(
    subject: string,
    body?: string,
    template?: string,
    data?: any,
    to?: string | string[],
    cc?: string | string[],
    bcc?: string | string[],
    attachments?: IMailAttachment[]
  ): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      /* Check mail service settings */
      if (!this._service || this._service.length <= 0) {
        return reject({
          boError: ERRORS.INVALID_MAIL_SERVER,
          boStatus: HTTP_STATUS.HTTP_4XX_NOT_ACCEPTABLE
        });
      }

      /* Check the mail receiver */
      if (
        (!to || (Array.isArray(to) && to.length === 0)) &&
        (!cc || (Array.isArray(cc) && cc.length === 0)) &&
        (!bcc || (Array.isArray(bcc) && bcc.length === 0))
      ) {
        return reject({
          boError: ERRORS.INVALID_MAIL_DATA,
          boStatus: HTTP_STATUS.HTTP_4XX_NOT_ACCEPTABLE
        });
      }

      /* Prepare the mail body */
      const bodyObj: any = {
        subject: subject,
        payload: data,
        to: to,
        cc: cc,
        bcc: bcc,
        attachments: attachments
      };
      if (body) {
        bodyObj["body"] = body;
      } else if (template) {
        bodyObj["template"] = template;
      } else {
        return reject({
          boError: ERRORS.INVALID_MAIL_DATA,
          boStatus: HTTP_STATUS.HTTP_4XX_NOT_ACCEPTABLE
        });
      }

      /* Send the mail notification */
      axios
        .post(`${this._service}/v1/mail/send`, bodyObj, {
          headers: {
            Authorization: `Bearer ${AuthenticationCtrl.token}`
          }
        })
        .then((response: AxiosResponse) => {
          try {
            /* Try to convert response body to JSON */
            JSON.parse(response.data);
            resolve();
          } catch {
            reject({
              boError: ERRORS.UNKNOWN_MAIL_SERVER_ERROR,
              boStatus: HTTP_STATUS.HTTP_4XX_FORBIDDEN
            });
            return;
          }
        })
        .catch((err: AxiosError) => {
          if (Objects.get(err, "response.data.error", null)) {
            return reject({
              boStatus: err.response.status,
              boError: {
                value: Objects.get(err, "response.data.error", SERVER_ERRORS.UNKNOWN_ERROR.value),
                str: Objects.get(err, "response.data.description", SERVER_ERRORS.UNKNOWN_ERROR.str)
              },
              boData: Objects.get(err, "response.data.data", null)
            });
          }

          reject({
            boError: ERRORS.UNKNOWN_MAIL_SERVER_ERROR,
            boStatus: HTTP_STATUS.HTTP_4XX_FORBIDDEN
          });
        });
    });
  }

  public attachmentFromFile(filepath: string, filename?: string): IMailAttachment {
    const fileBuff = fs.readFileSync(filepath);
    return {
      filename: filename || basename(filepath),
      content: fileBuff.toString("base64"),
      encoding: "base64"
    };
  }
}

export const MailCtrl: Mails = Mails.shared;


src/constants/errors.enum.ts
/**
 * Copyright (C) 2020-2022 ECUALEAD
 * All Rights Reserved
 * Author: Reinier Millo Sánchez <rmillo@ecualead.com>
 *
 * This file is part of the ECUALEAD Message Server API.
 * It can't be copied and/or distributed without the express
 * permission of the author.
 */

/**
 * Predefined errors
 */
export const ERRORS = {
  UNKNOWN_MAIL_SERVER_ERROR: {
    value: 1500,
    str: "unknown-mail-server-error"
  },
  NOT_AUTHORIZED: {
    value: 1501,
    str: "not-authorized"
  },
  INVALID_MAIL_TEMPLATE: {
    value: 1520,
    str: "invalid-mail-template"
  },
  INVALID_MAIL_SETTINGS: {
    value: 1521,
    str: "invalid-mail-settings"
  },
  INVALID_MAIL_SERVER: {
    value: 1522,
    str: "invalid-mail-server"
  },
  INVALID_MAIL_DATA: {
    value: 1523,
    str: "invalid-mail-data"
  },
  MAIL_SERVER_ERROR: {
    value: 1524,
    str: "mail-server-error"
  },
};
